File: python/agentbay/context_manager.py
from typing import Optional, List, Dict, Any, Callable
from agentbay.api.models import GetContextInfoRequest, SyncContextRequest
from agentbay.model.response import ApiResponse, extract_request_id
from .logger import get_logger, _log_api_call, _log_api_response, _log_api_response_with_details
import json
import time
import threading
import asyncio

# Initialize logger for this module
_logger = get_logger("context_manager")


class ContextStatusData:
    def __init__(
        self,
        context_id: str = "",
        path: str = "",
        error_message: str = "",
        status: str = "",
        start_time: int = 0,
        finish_time: int = 0,
        task_type: str = "",
    ):
        self.context_id = context_id
        self.path = path
        self.error_message = error_message
        self.status = status
        self.start_time = start_time
        self.finish_time = finish_time
        self.task_type = task_type

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> "ContextStatusData":
        return cls(
            context_id=data.get("contextId", ""),
            path=data.get("path", ""),
            error_message=data.get("errorMessage", ""),
            status=data.get("status", ""),
            start_time=data.get("startTime", 0),
            finish_time=data.get("finishTime", 0),
            task_type=data.get("taskType", ""),
        )


class ContextInfoResult(ApiResponse):
    def __init__(
        self,
        request_id: str = "",
        success: bool = True,
        context_status_data: Optional[List[ContextStatusData]] = None,
        error_message: str = "",
    ):
        super().__init__(request_id)
        self.success = success
        self.context_status_data = context_status_data or []
        self.error_message = error_message


class ContextSyncResult(ApiResponse):
    def __init__(
        self, request_id: str = "", success: bool = False, error_message: str = ""
    ):
        super().__init__(request_id)
        self.success = success
        self.error_message = error_message


class ContextManager:
    """
    Manages context operations within a session in the AgentBay cloud environment.

    The ContextManager provides methods to get information about context synchronization
    status and to synchronize contexts with the session.

    Example:
        ```python
        result = agent_bay.create()
        session = result.session
        info_result = session.context.info()
        print(f"Found {len(info_result.context_status_data)} context items")
        session.delete()
        ```
    """
    def __init__(self, session):
        self.session = session

    def info(
        self,
        context_id: Optional[str] = None,
        path: Optional[str] = None,
        task_type: Optional[str] = None,
    ) -> ContextInfoResult:
        """
        Get information about context synchronization status.

        Args:
            context_id: Optional ID of the context to get information for
            path: Optional path where the context is mounted
            task_type: Optional type of task to get information for (e.g., "upload", "download")

        Returns:
            ContextInfoResult: Result object containing context status data and request ID

        Example:
            ```python
            result = agent_bay.create()
            session = result.session
            info_result = session.context.info()
            for item in info_result.context_status_data:
                print(f"Context {item.context_id}: {item.status}")
            session.delete()
            ```
        """
        request = GetContextInfoRequest(
            authorization=f"Bearer {self.session._get_api_key()}",
            session_id=self.session._get_session_id(),
        )
        if context_id:
            request.context_id = context_id
        if path:
            request.path = path
        if task_type:
            request.task_type = task_type
        _log_api_call(
            "GetContextInfo",
            f"SessionId={self.session._get_session_id()}, ContextId={context_id}, Path={path}, TaskType={task_type}",
        )
        response = self.session._get_client().get_context_info(request)

        # Extract request ID
        request_id = extract_request_id(response)
        response_map = response.to_map()

        context_status_data: List[ContextStatusData] = []

        if isinstance(response_map, dict):
            body = response_map.get("body", {})
            try:
                response_body = json.dumps(body, ensure_ascii=False, indent=2)
            except Exception:
                response_body = str(body)

            # Check for API-level errors
            if not body.get("Success", True) and body.get("Code"):
                code = body.get("Code", "Unknown")
                message = body.get("Message", "Unknown error")
                _log_api_response_with_details(
                    api_name="GetContextInfo",
                    request_id=request_id,
                    success=False,
                    full_response=response_body
                )
                return ContextInfoResult(
                    request_id=request_id,
                    success=False,
                    context_status_data=[],
                    error_message=f"[{code}] {message}",
                )

            data = body.get("Data", {})
            context_status_str = data.get("ContextStatus", "")

            # Parse the context status data
            if context_status_str:
                try:
                    # First, parse the outer array
                    status_items = json.loads(context_status_str)
                    for item in status_items:
                        if item.get("type") == "data":
                            # Parse the inner data string
                            data_items = json.loads(item.get("data", "[]"))
                            for data_item in data_items:
                                context_status_data.append(
                                    ContextStatusData.from_dict(data_item)
                                )
                except json.JSONDecodeError as e:
                    _logger.error(f"❌ Error parsing context status: {e}")
                except Exception as e:
                    _logger.error(f"❌ Unexpected error parsing context status: {e}")

        # Log successful context info retrieval
        _log_api_response_with_details(
            api_name="GetContextInfo",
            request_id=request_id,
            success=True,
            key_fields={
                "context_id": context_id,
                "status_count": len(context_status_data)
            }
        )

        return ContextInfoResult(
            request_id=request_id,
            success=True,
            context_status_data=context_status_data,
            error_message="",
        )

    async def sync(
        self,
        context_id: Optional[str] = None,
        path: Optional[str] = None,
        mode: Optional[str] = None,
        callback: Optional[Callable[[bool], None]] = None,
        max_retries: int = 150,
        retry_interval: int = 1500,
    ) -> ContextSyncResult:
        """
        Synchronize a context with the session.

        This method supports two modes:
        - Async mode (default): When called with await, it waits for the sync operation to complete
        - Callback mode: When a callback is provided, it returns immediately and calls the callback when complete

        Args:
            context_id: Optional ID of the context to synchronize
            path: Optional path where the context should be mounted
            mode: Optional synchronization mode (e.g., "upload", "download")
            callback: Optional callback function that receives success status. If provided, the method
                     runs in background and calls callback when complete
            max_retries: Maximum number of retries for polling completion status (default: 150)
            retry_interval: Milliseconds to wait between retries (default: 1500)

        Returns:
            ContextSyncResult: Result object containing success status and request ID

        Example:
            Async mode - waits for completion:
            ```python
            result = agent_bay.create()
            session = result.session
            context = agent_bay.context.get('my-context', True).context
            sync_result = await session.context.sync(context.id, "/mnt/data")
            print(f"Sync success: {sync_result.success}")
            session.delete()
            ```

            Callback mode - returns immediately:
            ```python
            result = agent_bay.create()
            session = result.session

            def on_complete(success: bool):
                print(f"Sync completed: {success}")

            context = agent_bay.context.get('my-context', True).context
            await session.context.sync(context.id, "/mnt/data", callback=on_complete)
            session.delete()
            ```
        """
        request = SyncContextRequest(
            authorization=f"Bearer {self.session._get_api_key()}",
            session_id=self.session._get_session_id(),
        )
        if context_id:
            request.context_id = context_id
        if path:
            request.path = path
        if mode:
            request.mode = mode
        _log_api_call(
            "SyncContext",
            f"SessionId={self.session._get_session_id()}, ContextId={context_id}, Path={path}, Mode={mode}",
        )
        response = self.session._get_client().sync_context(request)

        # Extract request ID
        request_id = extract_request_id(response)
        response_map = response.to_map()

        success = False  # Initialize success variable

        if isinstance(response_map, dict):
            body = response_map.get("body", {})
            try:
                response_body = json.dumps(body, ensure_ascii=False, indent=2)
            except Exception:
                response_body = str(body)

            # Check for API-level errors
            if not body.get("Success", True) and body.get("Code"):
                code = body.get("Code", "Unknown")
                message = body.get("Message", "Unknown error")
                _log_api_response_with_details(
                    api_name="SyncContext",
                    request_id=request_id,
                    success=False,
                    full_response=response_body
                )
                return ContextSyncResult(
                    request_id=request_id,
                    success=False,
                    error_message=f"[{code}] {message}",
                )

            success = body.get("Success", False)

        # Log successful sync context call
        if success:
            _log_api_response_with_details(
                api_name="SyncContext",
                request_id=request_id,
                success=True,
                key_fields={
                    "context_id": context_id,
                    "path": path or "default"
                }
            )

        # If callback is provided, start polling in background thread (sync mode)
        if callback is not None and success:
            # Check if we're in an event loop
            try:
                asyncio.get_running_loop()
                # We're in an event loop, start polling in background thread
                poll_thread = threading.Thread(
                    target=self._poll_for_completion,
                    args=(callback, context_id, path, max_retries, retry_interval),
                    daemon=True,
                )
                poll_thread.start()
                return ContextSyncResult(request_id=request_id, success=success)
            except RuntimeError:
                # No event loop running, start polling in background thread
                poll_thread = threading.Thread(
                    target=self._poll_for_completion,
                    args=(callback, context_id, path, max_retries, retry_interval),
                    daemon=True,
                )
                poll_thread.start()
                return ContextSyncResult(request_id=request_id, success=success)

        # If no callback, wait for completion (async mode)
        if success:
            final_success = await self._poll_for_completion_async(
                context_id, path, max_retries, retry_interval
            )
            return ContextSyncResult(request_id=request_id, success=final_success)

        return ContextSyncResult(request_id=request_id, success=success)

    def _poll_for_completion(
        self,
        callback: Callable[[bool], None],
        context_id: Optional[str] = None,
        path: Optional[str] = None,
        max_retries: int = 150,
        retry_interval: int = 1500,
    ) -> None:
        """
        Polls the info interface to check if sync is completed and calls callback.

        Args:
            callback: Callback function that receives success status
            context_id: ID of the context to check
            path: Path to check
            max_retries: Maximum number of retries
            retry_interval: Milliseconds to wait between retries
        """
        for retry in range(max_retries):
            try:
                # Get context status data
                info_result = self.info(context_id=context_id, path=path)

                # Check if all sync tasks are completed
                all_completed = True
                has_failure = False
                has_sync_tasks = False

                for item in info_result.context_status_data:
                    # We only care about sync tasks (upload/download)
                    if item.task_type not in ["upload", "download"]:
                        continue

                    has_sync_tasks = True
                    _logger.info(
                        f"🔄 Sync task {item.context_id} status: {item.status}, path: {item.path}"
                    )

                    if item.status not in ["Success", "Failed"]:
                        all_completed = False
                        break

                    if item.status == "Failed":
                        has_failure = True
                        _logger.error(
                            f"❌ Sync failed for context {item.context_id}: {item.error_message}"
                        )

                if all_completed or not has_sync_tasks:
                    # All tasks completed or no sync tasks found
                    if has_failure:
                        _logger.warning("Context sync completed with failures")
                        callback(False)
                    elif has_sync_tasks:
                        _logger.info("✅ Context sync completed successfully")
                        callback(True)
                    else:
                        _logger.info("ℹ️  No sync tasks found")
                        callback(True)
                    break

                _logger.info(
                    f"⏳ Waiting for context sync to complete, attempt {retry+1}/{max_retries}"
                )
                time.sleep(retry_interval / 1000.0)

            except Exception as e:
                _logger.error(
                    f"❌ Error checking context status on attempt {retry+1}: {e}"
                )
                time.sleep(retry_interval / 1000.0)

        # If we've exhausted all retries, call callback with failure
        if retry == max_retries - 1:
            _logger.error(
                f"❌ Context sync polling timed out after {max_retries} attempts"
            )
            callback(False)

    async def _poll_for_completion_async(
        self,
        context_id: Optional[str] = None,
        path: Optional[str] = None,
        max_retries: int = 150,
        retry_interval: int = 1500,
    ) -> bool:
        """
        Async version of polling for sync completion.

        Args:
            context_id: ID of the context to check
            path: Path to check
            max_retries: Maximum number of retries
            retry_interval: Milliseconds to wait between retries

        Returns:
            bool: True if sync completed successfully, False otherwise
        """
        for retry in range(max_retries):
            try:
                # Get context status data
                info_result = self.info(context_id=context_id, path=path)

                # Check if all sync tasks are completed
                all_completed = True
                has_failure = False
                has_sync_tasks = False

                for item in info_result.context_status_data:
                    # We only care about sync tasks (upload/download)
                    if item.task_type not in ["upload", "download"]:
                        continue

                    has_sync_tasks = True
                    _logger.info(
                        f"🔄 Sync task {item.context_id} status: {item.status}, path: {item.path}"
                    )

                    if item.status not in ["Success", "Failed"]:
                        all_completed = False
                        break

                    if item.status == "Failed":
                        has_failure = True
                        _logger.error(
                            f"❌ Sync failed for context {item.context_id}: {item.error_message}"
                        )

                if all_completed or not has_sync_tasks:
                    # All tasks completed or no sync tasks found
                    if has_failure:
                        _logger.warning("Context sync completed with failures")
                        return False
                    elif has_sync_tasks:
                        _logger.info("✅ Context sync completed successfully")
                        return True
                    else:
                        _logger.info("ℹ️  No sync tasks found")
                        return True

                _logger.info(
                    f"⏳ Waiting for context sync to complete, attempt {retry+1}/{max_retries}"
                )
                await asyncio.sleep(retry_interval / 1000.0)

            except Exception as e:
                _logger.error(
                    f"❌ Error checking context status on attempt {retry+1}: {e}"
                )
                await asyncio.sleep(retry_interval / 1000.0)

        # If we've exhausted all retries, return failure
        _logger.error(f"❌ Context sync polling timed out after {max_retries} attempts")
        return False

File: java/agentbay/src/main/java/com/aliyun/agentbay/context/ContextManager.java
package com.aliyun.agentbay.context;

import com.aliyun.agentbay.session.Session;
import com.aliyun.agentbay.util.ResponseUtil;
import com.aliyun.wuyingai20250506.models.*;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.function.Consumer;
import java.util.stream.Collectors;

public class ContextManager {
    private static final ObjectMapper objectMapper = new ObjectMapper();

    private final Session session;

    public ContextManager(Session session) {
        this.session = session;
    }

    /**
     * Get context info for the session.
     * 
     * @return ContextInfoResult containing status data
     */
    public ContextInfoResult info() {
        return info(null, null, null);
    }

    /**
     * Get information about context synchronization status.
     *
     * @param contextId Optional ID of the context to get information for
     * @param path Optional path where the context is mounted
     * @param taskType Optional type of task to get information for (e.g., "upload", "download")
     * @return ContextInfoResult Result object containing context status data and request ID
     */
    public ContextInfoResult info(String contextId, String path, String taskType) {
        try {
            GetContextInfoRequest request = new GetContextInfoRequest();
            request.setAuthorization("Bearer " + session.getApiKey());
            request.setSessionId(session.getSessionId());

            if (contextId != null) {
                request.setContextId(contextId);
            }
            if (path != null) {
                request.setPath(path);
            }
            if (taskType != null) {
                request.setTaskType(taskType);
            }
            GetContextInfoResponse response = session.getAgentBay().getClient().getContextInfo(request);

            String requestId = ResponseUtil.extractRequestId(response);
            List<ContextStatusData> contextStatusData = new ArrayList<>();

            if (response != null && response.getBody() != null) {
                GetContextInfoResponseBody body = response.getBody();
                
                // Check for API-level errors
                Boolean bodySuccess = body.getSuccess();
                if (bodySuccess != null && !bodySuccess && body.getCode() != null) {
                    String code = body.getCode();
                    String message = body.getMessage() != null ? body.getMessage() : "Unknown error";
                    String errorMessage = String.format("[%s] %s", code, message);
                    return new ContextInfoResult(requestId, false, contextStatusData, errorMessage);
                }
                
                if (body.getData() != null) {
                    String contextStatus = body.getData().getContextStatus();
                    if (contextStatus != null && !contextStatus.isEmpty()) {
                        try {
                            // Parse the context status JSON
                            JsonNode statusItems = objectMapper.readTree(contextStatus);
                            for (JsonNode item : statusItems) {
                                if ("data".equals(item.get("type").asText())) {
                                    String dataString = item.get("data").asText();
                                    JsonNode dataItems = objectMapper.readTree(dataString);
                                    for (JsonNode dataItem : dataItems) {
                                        Map<String, Object> itemMap = objectMapper.convertValue(dataItem, Map.class);
                                        contextStatusData.add(ContextStatusData.fromMap(itemMap));
                                    }
                                }
                            }
                        } catch (Exception e) {
                        }
                    }
                }
            }

            return new ContextInfoResult(requestId, true, contextStatusData, "");

        } catch (Exception e) {
            return new ContextInfoResult("", false, new ArrayList<>(), "Failed to get context info: " + e.getMessage());
        }
    }

    /**
     * Sync context data (trigger upload).
     * 
     * @return ContextSyncResult indicating success/failure
     */
    public ContextSyncResult sync() {
        return sync(null, null, null);
    }

    /**
     * Synchronize a context with the session.
     *
     * @param contextId Optional ID of the context to synchronize
     * @param path Optional path where the context should be mounted; not limited to
     *             the path specified when creating the session (other backend-allowed
     *             paths are acceptable)
     * @param mode Optional synchronization mode (e.g., "upload", "download")
     * @return ContextSyncResult Result object containing success status and request ID
     */
    public ContextSyncResult sync(String contextId, String path, String mode) {
        try {
            SyncContextRequest request = new SyncContextRequest();
            request.setAuthorization("Bearer " + session.getApiKey());
            request.setSessionId(session.getSessionId());

            if (contextId != null) {
                request.setContextId(contextId);
            }
            if (path != null) {
                request.setPath(path);
            }
            if (mode != null) {
                request.setMode(mode);
            }
            SyncContextResponse response = session.getAgentBay().getClient().syncContext(request);

            String requestId = ResponseUtil.extractRequestId(response);
            boolean success = false;
            String errorMessage = "";

            if (response != null && response.getBody() != null) {
                SyncContextResponseBody body = response.getBody();
                success = body.getSuccess() != null ? body.getSuccess() : false;
                
                // Check for API-level errors
                if (!success && body.getCode() != null) {
                    String code = body.getCode();
                    String message = body.getMessage() != null ? body.getMessage() : "Unknown error";
                    errorMessage = String.format("[%s] %s", code, message);
                }
            }

            return new ContextSyncResult(requestId, success, errorMessage);

        } catch (Exception e) {
            return new ContextSyncResult("", false, "Failed to sync context: " + e.getMessage());
        }
    }

    /**
     * Sync context data with callback mode (non-blocking).
     * Returns immediately and calls the callback when sync completes.
     *
     * @param callback Callback function that receives success status (true if successful, false otherwise)
     * @return ContextSyncResult indicating initial sync trigger success/failure
     */
    public ContextSyncResult sync(Consumer<Boolean> callback) {
        return sync(null, null, null, callback);
    }

    /**
     * Sync context data with optional parameters and callback mode (non-blocking).
     * Returns immediately and calls the callback when sync completes.
     *
     * @param contextId Optional ID of the context to synchronize
     * @param path Optional path where the context should be mounted
     * @param mode Optional synchronization mode (e.g., "upload", "download")
     * @param callback Callback function that receives success status (true if successful, false otherwise)
     * @return ContextSyncResult indicating initial sync trigger success/failure
     */
    public ContextSyncResult sync(String contextId, String path, String mode, Consumer<Boolean> callback) {
        return sync(contextId, path, mode, callback, 150, 1500);
    }

    /**
     * Sync context data with optional parameters and callback mode (non-blocking).
     * Returns immediately and calls the callback when sync completes.
     *
     * @param contextId Optional ID of the context to synchronize
     * @param path Optional path where the context should be mounted
     * @param mode Optional synchronization mode (e.g., "upload", "download")
     * @param callback Callback function that receives success status (true if successful, false otherwise)
     * @param maxRetries Maximum number of retries for polling completion status (default: 150)
     * @param retryInterval Milliseconds to wait between retries (default: 1500)
     * @return ContextSyncResult indicating initial sync trigger success/failure
     */
    public ContextSyncResult sync(String contextId, String path, String mode, Consumer<Boolean> callback,
                                  int maxRetries, int retryInterval) {
        // First trigger the sync
        ContextSyncResult syncResult = sync(contextId, path, mode);

        if (!syncResult.isSuccess()) {
            // If sync trigger failed, call callback with false
            if (callback != null) {
                callback.accept(false);
            }
            return syncResult;
        }

        // Start polling in background thread
        if (callback != null) {
            Thread pollThread = new Thread(() -> {
                pollForCompletion(callback, contextId, path, maxRetries, retryInterval);
            });
            pollThread.setDaemon(true);
            pollThread.start();
        }

        return syncResult;
    }

    /**
     * Polls the info interface to check if sync is completed and calls callback.
     * 
     * <p>This method continuously polls the context status until all sync tasks
     * (upload/download) are completed or failed, or until the maximum number of
     * retries is reached.</p>
     *
     * @param callback Callback function that receives success status
     * @param contextId ID of the context to check
     * @param path Path to check
     * @param maxRetries Maximum number of retries
     * @param retryInterval Milliseconds to wait between retries
     */
    private void pollForCompletion(Consumer<Boolean> callback, String contextId, String path,
                                   int maxRetries, int retryInterval) {
        for (int retry = 0; retry < maxRetries; retry++) {
            try {
                // Get context status data
                ContextInfoResult infoResult = info(contextId, path, null);

                // Check if all sync tasks are completed
                boolean allCompleted = true;
                boolean hasFailure = false;
                boolean hasSyncTasks = false;

                for (ContextStatusData item : infoResult.getContextStatusData()) {
                    // We only care about sync tasks (upload/download)
                    String taskType = item.getTaskType();
                    if (taskType == null || (!"upload".equals(taskType) && !"download".equals(taskType))) {
                        continue;
                    }

                    hasSyncTasks = true;
                    if (!"Success".equals(item.getStatus()) && !"Failed".equals(item.getStatus())) {
                        allCompleted = false;
                        break;
                    }

                    if ("Failed".equals(item.getStatus())) {
                        hasFailure = true;
                    }
                }

                if (allCompleted || !hasSyncTasks) {
                    // All tasks completed or no sync tasks found
                    if (hasFailure) {
                        callback.accept(false);
                    } else if (hasSyncTasks) {
                        callback.accept(true);
                    } else {
                        callback.accept(true);
                    }
                    return;
                }
                Thread.sleep(retryInterval);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                callback.accept(false);
                return;
            } catch (Exception e) {
                try {
                    Thread.sleep(retryInterval);
                } catch (InterruptedException ie) {
                    Thread.currentThread().interrupt();
                    callback.accept(false);
                    return;
                }
            }
        }

        // Timeout
        callback.accept(false);
    }

    /**
     * Sync context data and wait for completion
     *
     * @return ContextSyncResult indicating success/failure after waiting for completion
     */
    public ContextSyncResult syncAndWait() {
        return syncAndWait(null, null, null);
    }

    /**
     * Sync context data with optional parameters and wait for completion
     *
     * @param contextId Context ID (optional)
     * @param path Path (optional)
     * @param mode Sync mode (optional)
     * @return ContextSyncResult indicating success/failure after waiting for completion
     */
    public ContextSyncResult syncAndWait(String contextId, String path, String mode) {
        return syncAndWait(contextId, path, mode, 150, 1500);
    }

    /**
     * Sync context data with optional parameters and wait for completion
     *
     * @param contextId Context ID (optional)
     * @param path Path (optional)
     * @param mode Sync mode (optional)
     * @param maxRetries Maximum number of retries for polling completion status (default: 150)
     * @param retryInterval Milliseconds to wait between retries (default: 2000)
     * @return ContextSyncResult indicating success/failure after waiting for completion
     */
    public ContextSyncResult syncAndWait(String contextId, String path, String mode, 
                                         int maxRetries, int retryInterval) {
        // First trigger the sync
        ContextSyncResult syncResult = sync(contextId, path, mode);
        
        if (!syncResult.isSuccess()) {
            return syncResult;
        }

        // Wait for sync to complete
        for (int retry = 0; retry < maxRetries; retry++) {
            try {
                // Get context status data
                ContextInfoResult infoResult = info(contextId, path, null);

                // Check if all sync tasks are completed
                boolean allCompleted = true;
                boolean hasFailure = false;
                boolean hasSyncTasks = false;

                for (ContextStatusData item : infoResult.getContextStatusData()) {
                    // We only care about sync tasks (upload/download)
                    String taskType = item.getTaskType();
                    if (taskType == null || (!"upload".equals(taskType) && !"download".equals(taskType))) {
                        continue;
                    }

                    hasSyncTasks = true;
                    if (!"Success".equals(item.getStatus()) && !"Failed".equals(item.getStatus())) {
                        allCompleted = false;
                        break;
                    }

                    if ("Failed".equals(item.getStatus())) {
                        hasFailure = true;
                    }
                }

                if (allCompleted || !hasSyncTasks) {
                    // All tasks completed or no sync tasks found
                    if (hasFailure) {
                        return new ContextSyncResult(syncResult.getRequestId(), false);
                    } else if (hasSyncTasks) {
                        return new ContextSyncResult(syncResult.getRequestId(), true);
                    } else {
                        return new ContextSyncResult(syncResult.getRequestId(), true);
                    }
                }
                Thread.sleep(retryInterval);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                return new ContextSyncResult(syncResult.getRequestId(), false);
            } catch (Exception e) {
                try {
                    Thread.sleep(retryInterval);
                } catch (InterruptedException ie) {
                    Thread.currentThread().interrupt();
                    return new ContextSyncResult(syncResult.getRequestId(), false);
                }
            }
        }

        // Timeout
        return new ContextSyncResult(syncResult.getRequestId(), false);
    }

    /**
     * Dynamically binds one or more contexts to the current session.
     *
     * <pre>{@code
     * ContextSync cs = ContextSync.create(contextId, "/tmp/ctx-data", null);
     * ContextBindResult result = session.getContext().bind(cs);
     * System.out.println("Bind success: " + result.isSuccess());
     * }</pre>
     *
     * @param contexts List of ContextSync objects to bind
     * @param waitForCompletion Whether to poll until all bindings are confirmed
     * @return ContextBindResult with the result of the operation
     */
    public ContextBindResult bind(List<ContextSync> contexts, boolean waitForCompletion) {
        if (contexts == null || contexts.isEmpty()) {
            return new ContextBindResult("", false, "At least one context is required");
        }

        try {
            List<BindContextsRequest.BindContextsRequestPersistenceDataList> persistenceDataList = new ArrayList<>();
            for (ContextSync ctx : contexts) {
                BindContextsRequest.BindContextsRequestPersistenceDataList item =
                    new BindContextsRequest.BindContextsRequestPersistenceDataList();
                item.setContextId(ctx.getContextId());
                item.setPath(ctx.getPath());
                if (ctx.getPolicy() != null) {
                    item.setPolicy(objectMapper.writeValueAsString(ctx.getPolicy()));
                }
                persistenceDataList.add(item);
            }

            BindContextsRequest request = new BindContextsRequest();
            request.setAuthorization("Bearer " + session.getApiKey());
            request.setSessionId(session.getSessionId());
            request.setPersistenceDataList(persistenceDataList);

            BindContextsResponse response = session.getAgentBay().getClient().bindContexts(request);
            String requestId = ResponseUtil.extractRequestId(response);

            if (response != null && response.getBody() != null) {
                BindContextsResponseBody body = response.getBody();
                if (body.getSuccess() != null && !body.getSuccess()) {
                    String code = body.getCode() != null ? body.getCode() : "Unknown";
                    String message = body.getMessage() != null ? body.getMessage() : "Unknown error";
                    return new ContextBindResult(requestId, false, String.format("[%s] %s", code, message));
                }
            }

            if (waitForCompletion) {
                List<String> contextIds = contexts.stream()
                    .map(ContextSync::getContextId)
                    .collect(Collectors.toList());
                pollForBindCompletion(contextIds, 60, 2000);
            }

            return new ContextBindResult(requestId, true);
        } catch (Exception e) {
            return new ContextBindResult("", false, "Failed to bind contexts: " + e.getMessage());
        }
    }

    /**
     * Binds a single context to the current session.
     *
     * @param context The ContextSync object to bind
     * @return ContextBindResult with the result of the operation
     */
    public ContextBindResult bind(ContextSync context) {
        List<ContextSync> list = new ArrayList<>();
        list.add(context);
        return bind(list, true);
    }

    /**
     * Lists all context bindings for the current session.
     *
     * <pre>{@code
     * ContextBindingsResult result = session.getContext().listBindings();
     * for (ContextBinding b : result.getBindings()) {
     *     System.out.println("Context " + b.getContextId() + " at " + b.getPath());
     * }
     * }</pre>
     *
     * @return ContextBindingsResult with the list of bindings
     */
    public ContextBindingsResult listBindings() {
        try {
            DescribeSessionContextsRequest request = new DescribeSessionContextsRequest();
            request.setAuthorization("Bearer " + session.getApiKey());
            request.setSessionId(session.getSessionId());

            DescribeSessionContextsResponse response =
                session.getAgentBay().getClient().describeSessionContexts(request);
            String requestId = ResponseUtil.extractRequestId(response);

            if (response != null && response.getBody() != null) {
                DescribeSessionContextsResponseBody body = response.getBody();
                if (body.getSuccess() != null && !body.getSuccess()) {
                    String code = body.getCode() != null ? body.getCode() : "Unknown";
                    String message = body.getMessage() != null ? body.getMessage() : "Unknown error";
                    return new ContextBindingsResult(requestId, false, new ArrayList<>(),
                        String.format("[%s] %s", code, message));
                }

                List<ContextBinding> bindings = new ArrayList<>();
                if (body.getData() != null) {
                    for (DescribeSessionContextsResponseBody.DescribeSessionContextsResponseBodyData item : body.getData()) {
                        bindings.add(new ContextBinding(
                            item.getContextId(),
                            item.getContextName(),
                            item.getPath(),
                            item.getPolicy(),
                            item.getBindTime()
                        ));
                    }
                }

                return new ContextBindingsResult(requestId, true, bindings);
            }

            return new ContextBindingsResult(requestId != null ? requestId : "", false, new ArrayList<>(),
                "Empty response");
        } catch (Exception e) {
            return new ContextBindingsResult("", false, new ArrayList<>(),
                "Failed to list bindings: " + e.getMessage());
        }
    }

    private void pollForBindCompletion(List<String> expectedContextIds, int maxRetries, int retryIntervalMs) {
        for (int i = 0; i < maxRetries; i++) {
            try {
                ContextBindingsResult result = listBindings();
                if (result.isSuccess()) {
                    Set<String> boundIds = new HashSet<>();
                    for (ContextBinding b : result.getBindings()) {
                        boundIds.add(b.getContextId());
                    }
                    if (boundIds.containsAll(expectedContextIds)) {
                        return;
                    }
                }
                Thread.sleep(retryIntervalMs);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                return;
            }
        }
    }
}
